### PR TITLE
[Core] Resubmit #47871 & #48030 and Fix the Failed Test for Windows

### LINKED ISSUE #48037: CI test windows://python/ray/tests:test_metrics_head is consistently_failing
CI test **windows://python/ray/tests:test_metrics_head** is consistently_failing. Recent failures: 
	- https://buildkite.com/ray-project/postmerge/builds/6555#01929227-abab-4e0c-a1e5-ea72448e2373
	- https://buildkite.com/ray-project/postmerge/builds/6555#019291cb-a8a6-4ff2-9ec2-79d00f9f661a

DataCaseName-windows://python/ray/tests:test_metrics_head-END
Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by MengjinYan**
## Why are these changes needed?
* Resubmit #47871 and #48030 as they are reverted as release blocker
* Fixed the failed test in windows CI test. 
  * The test `test_metrics_folder_and_content` failed because of the different file separator between linux and Windows
  * The PR fix the issue by using `os.path.join` to generate the path for validation instead of directly using formatted string

## Related issue number
Closes #48037

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by aslonnie — COMMENTED** [review]
is this PR ready for review yet?

File: python/ray/dashboard/modules/metrics/grafana_dashboard_provisioning_template.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
-DASHBOARD_PROVISIONING_TEMPLATE = """
-apiVersion: 1
+# DASHBOARD_PROVISIONING_TEMPLATE = """
```
[the remark this answers is not in the document]
Comment (MengjinYan) [reply python/ray/dashboard/modules/metrics/grafana_dashboard_provisioning_template.py:1]: Thanks for pointing out! 

This file and `python/ray/dashboard/modules/metrics/grafana_datasource_template.py` should be deleted as I moved the content to the more general template file: `python/ray/dashboard/modules/metrics/templates.py`. Something might be wrong when I reapplying the patch from the previous PR #47871. I'll update the PR to remove the files.

**Comment by MengjinYan** [comment]
> is this PR ready for review yet?

I original changes are reverted because a windows test failed because of the change. I did the fix and now waiting for all the CI tests to pass. It will be ready for review after the CI tests passed.

**Comment by rynewang** [comment]
This is surprising since I thought Python can treat `/` as dir separator on Windows as well. But anyway LGTM

**Comment by MengjinYan** [comment]
> This is surprising since I thought Python can treat `/` as dir separator on Windows as well. But anyway LGTM

Thanks! I think there are 2 things we hit in the Windows test:
1. The temp path in Windows are different from that on linux. On Windows, it is "C:\Users\..." while on Linux, we are using "/tmp/ray", so we cannot assume the temp path to always be "/tmp/ray"
2. Originally in the test, I used formatted string to generate the path instead of `os.path`. This cause the mismatch of the separator.

**Review by jjyao — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]

File: python/ray/dashboard/modules/metrics/install_and_start_prometheus.py [hunk 2]
```diff
@@ -74,12 +74,13 @@ def install_prometheus(file_path):
...
-    # Currently, Ray never modifies this config file, so we can just use the
-    # hardcoded path. (It just copies it to a more user-friendly location, in
-    # MetricsHead._create_default_prometheus_configs.)
-    # However, if in the future Ray ever modifies this file at runtime, we'll
-    # need to use the user-friendly location instead, and reload the config
-    # file after it's updated by Ray.
+    # The function assumes the Ray cluster to be monitored by Prometheus uses the
+    # default configuration with "/tmp/ray" as the default root temporary directory.
+    #
+    # This is to support the `ray metrics launch-prometheus` command, when a ray cluster
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/install_and_start_prometheus.py:80]: ...when a Ray cluster...

File: python/ray/dashboard/modules/metrics/install_and_start_prometheus.py [hunk 3]
```diff
@@ -74,12 +74,13 @@ def install_prometheus(file_path):
...
+    # is not started yet and we have no way to get a `--temp-dir` anywhere. So we choose
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/install_and_start_prometheus.py:81]: hasn't started yet and the user doesn't have a way to.....

File: python/ray/dashboard/modules/metrics/metrics_head.py [hunk 4]
```diff
@@ -186,17 +194,34 @@ def _create_default_grafana_configs(self):
...
         Creates the grafana configurations that are by default provided by Ray.
         """
-        grafana_config_output_path = os.path.join(self._metrics_root, "grafana")
-
-        # Copy default grafana configurations
-        if os.path.exists(grafana_config_output_path):
-            shutil.rmtree(grafana_config_output_path)
-        os.makedirs(os.path.dirname(grafana_config_output_path), exist_ok=True)
-        shutil.copytree(GRAFANA_CONFIG_INPUT_PATH, grafana_config_output_path)
+        # Create grafana configuration folder
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/metrics_head.py:197]: Create Grafana configuration folder.

File: python/ray/dashboard/modules/metrics/metrics_head.py [hunk 5]
```diff
@@ -186,17 +194,34 @@ def _create_default_grafana_configs(self):
...
+        if os.path.exists(self._grafana_config_output_path):
+            shutil.rmtree(self._grafana_config_output_path)
+        os.makedirs(self._grafana_config_output_path, exist_ok=True)
+
+        # Overwrite grafana's configuration file
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/metrics_head.py:202]: verwrite Grafana's configuration file.

File: python/ray/dashboard/modules/metrics/metrics_head.py [hunk 6]
```diff
@@ -186,17 +194,34 @@ def _create_default_grafana_configs(self):
...
+            ),
+            "w",
+        ) as f:
+            f.write(
+                GRAFANA_INI_TEMPLATE.format(
+                    grafana_provisioning_folder=grafana_prov_folder_with_latest_session
+                )
+            )
 
         # Overwrite grafana's dashboard provisioning directory based on env var
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/metrics_head.py:222]: Overwrite Grafana's dashboard provisioning directory based on env var.

File: python/ray/dashboard/modules/metrics/templates.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+PROMETHEUS_YML_TEMPLATE = """# my global config
+global:
+  scrape_interval: 10s # Set the scrape interval to every 10 seconds. Default is every \
+1 minute.
+  evaluation_interval: 10s # Evaluate rules every 10 seconds. The default is every 1 \
+minute.
+  # scrape_timeout is set to the global default (10s).
+
+scrape_configs:
+# Scrape from each ray node as defined in the service_discovery.json provided by ray.
```
Comment (angelinalg) [inline python/ray/dashboard/modules/metrics/templates.py:44]: Ray node...provided by Ray.
(capitalize Ray)

File: python/ray/dashboard/modules/tests/test_metrics_integration.py [hunk 8]
```diff
@@ -38,5 +40,18 @@ def test_e2e(capsys):
     subprocess.run(["kill", str(pid)])
 
 
+def test_prometheus_config_content():
+    # Test to make sure the content in the hardcoded file
+    # (python/ray/dashboard/modules/metrics/export/prometheus/prometheus.yml) will
+    # always be the same as the template (templates.py) used to generate prometheus
```
Comment (angelinalg) [inline python/ray/dashboard/modules/tests/test_metrics_integration.py:46]: Prometheus
(capitalize Prometheus)

File: python/ray/dashboard/modules/tests/test_metrics_integration.py [hunk 9]
```diff
@@ -38,5 +40,18 @@ def test_e2e(capsys):
...
+    # config file when ray startup
```
Comment (angelinalg) [inline python/ray/dashboard/modules/tests/test_metrics_integration.py:47]: when Ray starts up.

**Review by angelinalg — COMMENTED** [review]
Some style nits if they can be fixed in a separate PR.

**Pull request merged**